mgr: Add get_rates_from_data to mgr_util.py

Devp00l opened the pull request:
The function calculates the difference between the values
in a time series list.

The previous location of it was in the dashboard, that's why the
function has doc tests. In order to be able to test them as part of the
manager utilities a bit of tox magic was added.

Fixes: https://tracker.ceph.com/issues/40365
Signed-off-by: Stephan Müller <smueller@suse.com>

LenzGr commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -1,6 +1,15 @@
 import socket
 
+try:
+    from more_itertools import pairwise
```
Does this lead to a new Python package dependency? Or are these modules part of Python core?

sebastian-philipp replied:
I'd prefer to have this as a real dependency in octopus, indeed.

Devp00l replied:
We used them in the dashboard so I assumed we could use them here too, without further declaration, am I wrong with this assumption?

sebastian-philipp replied:
right now, it doesn't add another hard dependency, no.

Devp00l commented:
I added a commit that should make tox fail if it is executed - ATM it's not failing.

@callithea Maybe you can help me to make "make check" fail as intended :)

Also I'm still not sure where to put the documentation :/

callithea commented:
If it's not part of any of the ``run-tox-`` scripts (see ``make check`` output) I'd guess you would need to add/register it somewhere first?

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -1,6 +1,15 @@
+    def pairwise(iterable):
+        from itertools import tee
+        a, b = tee(iterable)
+        next(b, None)
```
(I know this code snippet was already there, and that it's pretty straightforward, but...) this is what I meant the other day in the F2F: as we don't want to deal with the burden of packaging dependencies for all systems Ceph has to run into (because it's a real pain), we end up copy-pasting [recipes from some open source library](https://github.com/erikrose/more-itertools) and infringing its license (`The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.`, _yeah, this maybe is just a tiny portion_) [[more-itertools LICENSE]](https://github.com/erikrose/more-itertools/blob/93e50a42b9d7ae77595918d6951cd876fee92473/LICENSE#L10-L11).

I've had to do this too for `pluggy` (as I couldn't find packages for them in Trusty/CentOS7/...), but I'm not proud of it :sweat_smile: Don't get me wrong, I really think reusing code is the right approach, but we shouldn't do it surreptitiously. IMHO we should face it openly and discuss how to do it, rather than leaving it up to every developer good will.

IMHO this is not better than distributing a dependency outside the distro package manager (e.g.: via pip). Is anyone tracking this code snippet for vulnerabilities reported in the source package? I'm afraid not. It requires manual review and update. Could we benefit from reusing other open source libraries? I think we could and should.

sebastian-philipp replied:
We could make this a hard dependency, yes.

epuertat replied:
Given there's no source for this package in CentOS/RHEL (AFAICT), and we haven't agreed on another solution to increase usage of existing Python libraries (in the same way we do for NPM packages), let's go with the current approach (I'd prepend the copyright notice, just to comply with MIT license).

sebastian-philipp replied:
fyi, this `pairwise` originates from `https://docs.python.org/3/library/itertools.html#itertools-recipes` and not from more-itertools

Devp00l commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -106,3 +119,34 @@ def is_ipv6_enabled():
+    [(2, 0.5), (3, 1.0), (4, 0.0)]
+    """
+    if not data:
+        return [(0, 0.0)]
```
I fully agree, we should just return an empty array and don't create something that doesn't exist. But it was implemented in this function before. I just look add the commit description of 58ffd14 introduced by @b-ranto, he added the two conditionals in order to fix API test issues that randomly occurred. He introduced that a year ago (29.05.2018), so I'm not sure we really need it anymore.

sebastian-philipp replied:
Returning an empty list should be find.

Devp00l commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -106,3 +119,34 @@ def is_ipv6_enabled():
+    if not data:
+        return [(0, 0.0)]
+    if len(data) == 1:
+        return [(data[0][0], 0.0)]
```
From the same commit as mentioned above ;)

sebastian-philipp replied:
same here

Devp00l commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -106,3 +119,34 @@ def is_ipv6_enabled():
+    return [(data2[0], differentiate(data1, data2)) for data1, data2 in pairwise(data)]
+
+
+def differentiate(data1, data2):
```
That's not true, if you reduce PGs of a pool for example the read / write ops and bytes decrease without any overflow - that was the reason to use the absolute value instead. (But this could also be a bug in ceph)

rjfd replied:
I don't understand one thing, why can't we return a negative number here. If the derivative function returns a negative value it means that something is decreasing. I'm afraid by taking the absolute value we are returning incorrect data.

Devp00l replied:
One segment of a time series data often conists of the following:
`[tiemstamp, sum of something]` but this leaves the question open how should our rates look like?
In the end we want to show a graph that clearly shows if something is going on.

For example if we decrease the pgs of a pool with some data on it the time series sum for writing bytes and operation will drop over time.
As our writing speed and writing operations cannot grow negative we should show the values positive, as this would distract the user and would mess up the current graph.
To fix the graph we would have to center zero leaving us with half the space to display positive values for this rare case of negative values.

Nevertheless it could also be a bug in the time series usage for this particular case.

rjfd replied:
> For example if we decrease the pgs of a pool with some data on it the time series sum for writing bytes and operation will drop over time.

Maybe this is the problem. Why would the value of written bytes decrease when the PG number decreases? Can you reproduce this behavior?

Nevertheless, we shouldn't use  `abs()` because from a mathematical perspective is just wrong.

Devp00l replied:
Yes I can reproduce this, @callithea wrote me a [comment](https://github.com/ceph/ceph/pull/28153#issuecomment-493469846) about this on an already merged PR where I already introduced the use of `abs`.

Mathematical it's presenting the absolute difference.
Using just the difference, would cause problems if the data we are receiving is not to 100% accurate. With decreasing sums and duplicate timestamps, we don't have that yet IMO.

In the end we don't want our users to freak out because their writing speed is negative ;), as it's like driving backwards with the car and it shows -10 km/h on the tachometer :smile:

rjfd replied:
> Yes I can reproduce this, @callithea wrote me a [comment](https://github.com/ceph/ceph/pull/28153#issuecomment-493469846) about this on an already merged PR where I already introduced the use of `abs`.
> 
> Mathematical it's presenting the absolute difference.
> Using just the difference, would cause problems if the data we are receiving is not to 100% accurate. With decreasing sums and duplicate timestamps, we don't have that yet IMO.
> 
> In the end we don't want our users to freak out because their writing speed is negative ;), as it's like driving backwards with the car and it shows -10 km/h on the tachometer

That's why we should fix the root cause of the problem and not change the derivative function definition. We must understand why the total number of write/read bytes decreases when the PG number decreases.

epuertat replied:
I'm with @rjfd here. If this all comes from a cumulative perf counter (read/written bytes to a pool), it should never decrease, unless that counter exhibits some weird semantics (on PG shrink written bytes are proportionally shrunk?). In any case that would smell like something's happening there :question:

sebastian-philipp commented:
Fascinating that we're reviewing code that was not really modified by this PR :-P

Devp00l commented:
jenkins test make check

Devp00l commented:
jenkins test make check

epuertat commented on src/pybind/mgr/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+pytest-cov==2.5.1
+py==1.5.2
+six==1.11.0
+tox==2.9.1
```
Other commit jumped into this PR?

Devp00l replied:
No that's needed ;)

Devp00l commented:
Addressed all comments :)
This PR is no draft anymore.
I added `HACKING.md` inside the manager directory with instructions on how to use the tests.
I used `mark down` instead of `rest` as `Github` seems to stopped supporting it.

epuertat commented on src/pybind/mgr/plugins/more_itertools.py:
```diff
@@ -0,0 +1,27 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
```
Thanks! BTW, I'd suggest to move this out of `plugins` dir (I have a PR to remove lru/ttl caches from there too). As other modules may benefit from this, why not `src/pybind/mgr/utils`?

sebastian-philipp replied:
See above.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -164,3 +166,65 @@ def verify_tls_files(cert_fname, pkey_fname):
+    0.5
+    """
+    time_diff = float(p2[0] - p1[0])
+    deriv = (p2[1] - p1[1]) / (time_diff if time_diff != 0 else 1)
```
I still don't understand why we assume 1 second/time-unit interval when `t0 == t1`?

If that ever happens, let's either drop that sample, so no rate at all, or trigger an Exception, so the outer caller knows this weird condition happened.

Devp00l replied:
We don't. The reason to use one here is purely mathematically as we can't divide through zero.
If the value still is the same nothing changed through the two samples.

epuertat replied:
That's wrong, and I'm sorry to be that picky with this code, but when you implement a math operation you cannot be creative. Imagine that some creative Python core developer would have decided that, to save you from the pain of dealing with exceptions, when you divide by 0, instead of raising `ZeroDivisionError`, Python would do the division by 1. Python would be the most unreliable programming language ever.

Any sample with the same timestamp should be considered an invalid one, and we should trigger an exception as soon we have them (before spreading them into other methods).

Why? Here there are 2 possible inputs when `time_diff = 0` and these are the outputs according to this code:

- `p2[1] == p1[1]`: then `rate = 0`. 
  - That basically means a dashboard operator will see a notorious drop to zero in whatever graph they're watching at, and if they're serious about their jobs they'll immediately start an investigation to find why this is happening: network packet drops? Hard-disk read retries?

- `p2[1] != p1[1]`: then `rate = X`. 
  - How do you decide whether rate is `X` or `-X`, since both samples belong to the same timestamp? Is it an increasing or a decreasing rate?

Devp00l replied:
I see your point, and we already discussed that in the other PR that was closed, that it's possible for some unknown reason...

The position in the time series data would determine who was recorded first.

As the time series data contains multiple segments I would not want to raise an error because of a minor issue that's not as tragic. IMO if we can predict it and know how to work around it we should do that instead of raising errors for a bad user experience.

How about we just drop the segment?

After reading your comment this sounds like the best idea IMO, we could also log the double entry to make sure it makes some noise even if it's little.

epuertat replied:
> As the time series data contains multiple segments I would not want to raise an error because of a minor issue that's not as tragic. IMO if we can predict it and know how to work around it we should do that instead of raising errors for a bad user experience.

Sorry I cannot agree with this: data can be valid or invalid. You cannot let some data pass because it's not _that invalid_.

- Duplicate samples is valid data.
  - Returning rate 0 for them is **invalid data**.
- Mismatching samples for the same timestamp is **invalid data**.
  - Inferring any rate from them is **invalid data**.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -164,3 +166,65 @@ def verify_tls_files(cert_fname, pkey_fname):
+    -0.5
+    >>> derivative(*zip(times, values), True)
+    0.5
+    """
```
Nice unit test coverage. What about adding edge cases:
- `t=[0,0]`, `v=[0,0]` -> ZeroDivisionError -> NotEnoughTimeHasPassedError?
- `t=[0,0]`, `v=[0, float('nan')]` -> nan 
- `t=[0,float('inf')]`, `v=[0, 3]` -> 0
- `t=[0,float('inf')]`, `v=[0, float('inf')]` -> nan

sebastian-philipp replied:
I'd prefer to have the scope of the doc tests limited. Doc tests are not the best place for extensive edge case testing.

epuertat replied:
Why not? For simple functions, like math operations, they become both documentation and tests. And edge cases are the things you definitely want to include in a unit test, because otherwise you may have undefined behaviours.

sebastian-philipp replied:
> Why not?

Because verbosely documenting edge cases rarely improve the readability of docstrings. I'd argue for keeping the number of tests within docstrings limited to what actually shows non-trivial behaviour.

epuertat replied:
I personally find that doctests are documentation that becomes tested (for free). If @Devp00l adds a comment explaining this function's quirks: "when 2 samples refer to the same timestamp this returns the rate as if they were 1 second separated". If someone else later changes the behaviour and they somehow forget to amend that documentation accordingly (which is more common that it seems), there we have mismatching docs.

sebastian-philipp replied:
Just keep in mind the doc string is already extraordinary long. At some point we have to evaluate how much documentation we want for purely internal functions.

Devp00l replied:
I will only add the case for the same timestamp, which will return 0 and should not return a failure or just skip the value.

epuertat replied:
> I will only add the case for the same timestamp, which will return 0 and should not return a failure or just skip the value.

Then I'm afraid that I haven't made myself understood. The "same-timestamp-case" should never return 0, as `math.log(0)` never returns anything except `ValueError: math domain error`.

epuertat commented on src/pybind/mgr/dashboard/services/ceph_service.py:
```diff
@@ -187,7 +185,7 @@ def get_rate(cls, svc_type, svc_name, path):
 
         if data and len(data) > 1:
-            return differentiate(*data[-2:])
+            return derivative(*data[-2:])
```
That syntax is totally ok, but might require some extra attention (and if someone tries with 3 samples `derivative(*data[-3:])`, it'd still work... but weirdly). So, what about?:
```suggestion
            return derivative(data[-2], data[-1])
```

sebastian-philipp replied:
This seems like an improvement for a different PR to me?

epuertat replied:
It's just a change in the way the function is fed. I don't think it qualifies as an improvement.

sebastian-philipp replied:
As the API of this function stays unchanged, I'd argue for keeping the noise in this PR small.

epuertat replied:
The splat operator is a language feature for feeding functions with a variable number of arguments (or when there're plenty of them). As in this case we're already checking that the list will have at least 2 elements, I don't understand why we should do any trick.

sebastian-philipp replied:
We're dealing with the tail of a list aka `data[-2:]`, meaning it will obviously not properly work, if you call it with `reverse(data[-2:])`. We're not dealing with two individual parameters. I'm :+1: for keeping it this way.

Devp00l replied:
I will change the condition above to expect 2 elements in the list.

epuertat commented on src/pybind/mgr/dashboard/services/ceph_service.py:
```diff
@@ -114,14 +106,20 @@ def get_pool_list_with_stats(cls, application=None):
             def get_rate(series):
                 if len(series) >= 2:
-                    return differentiate(*list(series)[-2:])
+                    return derivative(*list(series)[-2:], True)
```
Same here?

Devp00l replied:
Here we already expect a list with 2 or more elements.

epuertat commented:
And, BTW, I've had to filter out 2 commits. They seem to come from the unit-test coverage refactoring.

sebastian-philipp commented on src/pybind/mgr/plugins/more_itertools.py:
```diff
@@ -0,0 +1,27 @@
+"""
+Copyright (c) 2012 Erik Rose
```
Ais I said, the copyright is wrong. it originates from https://docs.python.org/3/library/itertools.html#itertools-recipes instead.

Devp00l replied:
Removed the copyright

sebastian-philipp commented on src/pybind/mgr/HACKING.md:
```diff
@@ -0,0 +1,56 @@
+
+## Unit Testing
+
+In the manager our unit tests are based on **tox**.
```
Which is actually a bit misleading. We're only testing very few files, not this directory. Instead, this directory mostly consist of files that are tested by individual unit tests not covered here.

sebastian-philipp commented on src/pybind/mgr/run-tox.sh:
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
```
@tchaikov just removed all other `ron-tox.sh` files from the project, excep for `src/script/run_tox.sh`. Can you make use of the other one instead?

sebastian-philipp commented on src/pybind/mgr/HACKING.md:
```diff
@@ -0,0 +1,56 @@
+> Currently only [doctests](https://docs.python.org/3/library/doctest.html) inside
+> `mgr_util.py` are run
+
+### How to run the tests
```
I don't think we need this section anymore at this place. With the project global `src/script/run_tox.sh`, this could potentially be moved to `/doc/`. I for one would prefer to have no documentation here, instad of some outdated one.

sebastian-philipp commented on src/pybind/mgr/tox.ini:
```diff
@@ -0,0 +1,26 @@
+    cov:  UNITTEST = true
+    cov:  COVERAGE_FILE = .coverage.{envname}
+commands=
+    pip install -r {toxinidir}/requirements.txt
```
Would it be possible to use [`deps=`](https://tox.readthedocs.io/en/latest/config.html#conf-deps) here? always running `pip install` seems a bit heavy

sebastian-philipp commented:
@epuertat

We're not trying to move two simple functions to another location since **February**. We're now really wasting time here, take the rework that needs to be done in order to cope with @tchaikov 's (awesome) `run_tox.sh` unification as an example. 

That's why I'm recommending things to be stripping things from this PR. This PR already grew way too big for this small improvement.

epuertat commented:
> @epuertat We're now trying to move two simple functions to another location since **February**. We're now really wasting time here, take the rework that needs to be done in order to cope with @tchaikov 's (awesome) `run_tox.sh` unification as an example.
> 
> That's why I'm recommending things to be stripping things from this PR. This PR already grew way too big for this small improvement.

I dont' think this has intentionally grown. It looks like when pushed it got mixed with other commits. 

My comment (I'm not "requesting", as per Github's PR Review, any changes in my last review) is that `plugins/` is not the proper place for this to be (as that's meant for dashboard plugins, not utils).

sebastian-philipp commented:
> > @epuertat We're now trying to move two simple functions to another location since **February**. We're now really wasting time here, take the rework that needs to be done in order to cope with @tchaikov 's (awesome) `run_tox.sh` unification as an example.
> > That's why I'm recommending things to be stripping things from this PR. This PR already grew way too big for this small improvement.
> 
> I dont' think this has intentionally grown. It looks like when pushed it got mixed with other commits.
> 
> My comment (I'm not "requesting", as per Github's PR Review, any changes in my last review) is that `plugins/` is not the proper place for this to be (as that's meant for dashboard plugins, not utils).

Just saying that we're successfully bikeshedding since February.

epuertat commented:
> Just saying that we're successfully bikeshedding since February.

I partly agree. The amount of discussion for the amount of code/functionality brought by this PR is clearly an overkill, but I don't think it's exactly bikeshedding: a great part of the discussion involved understanding the underlying issue that this fix was trying to tackle, and in fact @rjfd caught and fixed an issue in ceph-mgr thanks to it.

This PR specifically dates from July. I know the fix has changed hands, but the issue seem to remain almost the same since #26270 and #28153, at least from my point of view:
- straightforward `derivative` method with no custom magic inside: `(f[t_1] - f[t_0]) / (t_1 - t_0)`

Anything else from my side (file location, doctest, etc.) are suggestions. Nevertheless, there are other 6 reviewers summoned in this PR, so let them speak out.

Devp00l commented:
I'll address all the comments soon. Thanks for your input.

Devp00l commented:
Addressed nearly all comments - Please click resolve for that that I have addressed in your opinion.

Note I added the documentation to `/doc`, please review it.

epuertat commented on src/pybind/mgr/dashboard/services/ceph_service.py:
```diff
@@ -114,14 +106,20 @@ def get_pool_list_with_stats(cls, application=None):
 
             for stat_name, stat_series in stats.items():
+                # Using the absolute value for 'rate' and 'rates' is a temporary workaround for
+                # the pg decrease issue. A decrease will reduce the maximum amount in all time
```
Could you please add here the URL to the issue? I couldn't find it.

I do think we must not sweep issues under the rug. If we are facing duplicate samples, mismatching samples for the same timestamp, or counters that are expected to behave monotonically but somehow decrease, we shouldn't make it up. This is the proof that Ceph-Dashboard is doing a great job exposing issues.

So I still think we should remove the `abs()` (now it can be optionally enabled... but it's enabled most of the times!) and the ` deriv = (p2[1] - p1[1]) / (time_diff if time_diff != 0 else 1)` should be just `deriv = (p2[1] - p1[1]) / (p2[0] - p1[0])`.

tchaikov commented:
retest this please.

Devp00l commented:
I have to make some changes, as I just found out a new inconsistency. Some time series data provides the time in milliseconds (osds) (IMO milliseconds should be common) and some provide the time in seconds (pools).

Let's first assume ms is common, than we should show what happened in  a second and not a ms so we have to multiply the result with 1k. 

I also put the rate function one level up as it uses the same data structure.

I added a private filter function where the data will be filtered for valid segments.

All these changes are not submitted yet, but will come soon. But we can discuss the time issue until than ;) (Best discussed in the stand up)

epuertat commented:
@Devp00l @rjfd and I had a face to face discussion on this and we all agreed on:
- `get_most_recent_rate` will return `None` when the rate list is empty.
- `_filter_time_series` logic is ok as it is (though @rjfd mentioned, as I did, that it's not easy to understand the list comprehension).
- Duplicate samples will be logged.
- `abs()` won't be applied to rates. If a negative rate appears, when not expected (as it happens with IOPS-related counters related on PG merge scenarios), we'll leave a negative rate and report the issue back to the corresponding Ceph core team.

Devp00l commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -176,3 +178,124 @@ def verify_tls_files(cert_fname, pkey_fname):
+    >>> _filter_time_series([[1, 42]])
+    []
+    >>> _filter_time_series([[0, 2], [0, 2]])
+    []
```
I amended the code so it shows the debug output, now it's easier to spot which of the two numbers is the timestamp and what is the value.

The timestamp is the first number.

As said during our discussion today, one will be filtered one sample will be left and if only one sample exists `None` will be returned.

Devp00l commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -176,3 +178,124 @@ def verify_tls_files(cert_fname, pkey_fname):
+    """
+    if data is None:
+        return []
+    max = len(data)-1
```
I agree the naming was pretty badly chosen, now it's `last_index`.

Devp00l commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -176,3 +178,124 @@ def verify_tls_files(cert_fname, pkey_fname):
+    if data is None:
+        return []
+    max = len(data)-1
+    data = [d for i, d in enumerate(data) if i < max and d[0] != data[i+1][0] or i == max]
```
Yes it was a bit hard to read that's why I created a lambda which now is used inside the loop. This way it's easier to revert the outcome in case of a sample with the same timestamp as another.

Hope my solution suits you as well.

Devp00l commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -176,3 +178,124 @@ def verify_tls_files(cert_fname, pkey_fname):
+        return abs(deriv) if use_abs else deriv
+    except ZeroDivisionError as error:
+        logger.warning(
+            'Derivative was called with the same timestamp ({}, {})'.format(p1, p2))
```
Removed it.

rjfd commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -176,3 +178,109 @@ def verify_tls_files(cert_fname, pkey_fname):
+        logger.debug('Invalid time series samples {} in {}'.format(removed, data))
+    if len(filtered) < 2:
+        return None
+    return filtered
```
I think the current implementation is still confusing. What do you think about changing it to the following implementation?

```python
filtered = []
for i in range(len(data)-1):
  if data[i][0] == data[i+1][0]:
    log.debug("Found duplicate sample with same timestamp: %s, %s", data[i], data[i+1])
    continue
  filtered.append(data[i])
filtered.append(data[-1])
return filtered
```

Devp00l replied:
Thanks, nice work :)

rjfd commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -176,3 +178,109 @@ def verify_tls_files(cert_fname, pkey_fname):
+    It also enforces the list to contain at least two samples.
+
+    All filtered values will be shown in the debug log. If values were filtered it's a bug in the
+    time series data collector, please report it.
```
Can you describe the filtering rule in the comment? Example, "this method removes consecutive samples with the same timestamp"

Devp00l replied:
done

rjfd commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -176,3 +178,109 @@ def verify_tls_files(cert_fname, pkey_fname):
+    All filtered values will be shown in the debug log. If values were filtered it's a bug in the
+    time series data collector, please report it.
+
+    :param data: Time series data where the time is provided in seconds
```
Can you specify which value of the tuple corresponds to the timestamp and the value?

Devp00l replied:
done

Devp00l commented:
Addressed all comments

callithea commented:
@Devp00l: I noticed that changes have been pushed, so I removed the testing labels for now. Please go ahead and re-add the needs-qa label if your PR is ready to go through QA.

callithea commented:
jenkins test dashboard

callithea commented:
@tchaikov could you please take a look at the build errors?

tchaikov commented on src/pybind/mgr/utils/more_itertools.py:
```diff
@@ -0,0 +1,7 @@
+from itertools import tee
```
@callithea @Devp00l this is a new module. if we want to use it at runtime. we need to package it. but currently, neither ceph.spec.in nor debian/* packages this module.

i'd suggest promote this module and move it into `python-ceph-common` package.

@sebastian-philipp what do you think?

callithea replied:
@tchaikov thanks for your feedback!

sebastian-philipp replied:
+1

callithea commented:
QA run [WIP]: http://pulpito.ceph.com/laura-2019-11-06_14:52:29-rados:dashboard-wip-laura-testing-28603-11-06-distro-basic-smithi/

LenzGr commented:
What's the status of this PR? Will the test/build failures be addressed? What's missing in order to get this merged?

Devp00l commented:
The failure is not addressed yet.

Devp00l commented:
Removed the more_intertools integration and reimplemented the pairwise function as private function in mgr util as the pairwise function originates from the [python docs](https://docs.python.org/3/library/itertools.html#itertools-recipes)

This will fix all build failures.

epuertat commented on src/pybind/mgr/mgr_util.py:
```diff
@@ -180,3 +182,127 @@ def verify_tls_files(cert_fname, pkey_fname):
+    """
+    a, b = tee(iterable)
+    next(b, None)
+    return zip(a, b)
```
Please, check this ticket: https://tracker.ceph.com/issues/43055. I'd suggest you to replace this implementation with another one not relying on `tee()`.

Devp00l commented:
@callithea I wasn't able to recreate your failure, I just hope the rebase fixed it.

@epuertat I changed the pairwise function as suggested.

callithea commented:
> @callithea I wasn't able to recreate your failure, I just hope the rebase fixed it.
> 
> @epuertat I changed the pairwise function as suggested.

Unfortunately there are still `make check` errors:
```
1 - run-tox-mgr (Failed)
2 - run-tox-mgr-dashboard (Failed)
144 - unittest_chain_xattr (Timeout)
```

Devp00l commented:
jenkins test make check

Devp00l commented:
jenkins test dashboard

Devp00l commented:
jenkins test dashboard backend

Devp00l commented:
jenkins test make check arm64

smithfarm commented:
@Devp00l Unless you have a special interest in ARM, you needn't worry about the arm64 make check test - it is not mandatory, and is known to be unreliable.

s0nea commented:
jenkins test dashboard backend

callithea commented:
jenkins test dashboard

smithfarm commented:
@Devp00l If this PR will fix https://tracker.ceph.com/issues/43224 as well as https://tracker.ceph.com/issues/40365 does that mean one of those can be marked as a Duplicate of the other? Or do both need to stay open?

Devp00l commented:
@smithfarm I marked it as duplicate ;)

s0nea commented:
jenkins test make check

callithea commented:
jenkins test dashboard

Devp00l commented:
I changed the return type of no useful data to return the same type as with useful data, but in a vacuous state, as the dashboard failed on the creation of a pool because the `None` return.

This means if a list is the return type an empty list will be returned. 

In case of an float `0.0` will be returned which is fine as the float should resemble if the rate has gone up or done, but at the beginning of a rate (after creation), the created pool can have less than 2 timestamps for up to 2 seconds, which are not enough to declare a useful rate.

@epuertat I hope you are fine with that.

Devp00l commented:
jenkins test dashboard

Devp00l commented:
jenkins test docs

liewegas commented:
jenkins test docs

liewegas commented:
@callithea make check passing.. approve?

@epuertat are you ok with the current version?

Devp00l commented:
@epuertat I removed the unused import

Devp00l commented:
jenkins test dashboard

s0nea commented:
@liewegas or @tchaikov can you tell which QA suite do I need to run in order to get this PR tested? - The whole rados suite or is there a suitable subset?

liewegas commented:
just the rados/dashboard tests, i suspect, but i'll include this in my next rados batch.

s0nea commented:
> just the rados/dashboard tests, i suspect, but i'll include this in my next rados batch.

Okay, thank you!

liewegas commented:
tests pass!

liewegas commented:
I think @Devp00l deserves some sort of aware for this PR!
